File: 백준/Silver/15787. 기차가 어둠을 헤치고 은하수를/기차가 어둠을 헤치고 은하수를.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())

trains = [0 for _ in range(N)]

mask = (1 << 20) - 1

for _ in range(M):
    command = list(map(int, input().split()))

    i = command[1]

    match command[0]:
        case 1:
            x = command[2]
            trains[i - 1] |= 1 << (x - 1)
        case 2:
            x = command[2]
            trains[i - 1] &= ~(1 << (x - 1))
        case 3:
            trains[i - 1] <<= 1
        case 4:
            trains[i - 1] >>= 1
    
    trains[i - 1] &= mask

print(len(set(trains)))

File: 백준/Silver/15787. 기차가 어둠을 헤치고 은하수를/기차가 어둠을 헤치고 은하수를.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[] trains;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        trains = new int[N];

        int mask = (1 << 20) - 1;

        for (int idx = 0; idx < M; idx++) {
            st = new StringTokenizer(br.readLine());

            int c = Integer.parseInt(st.nextToken());
            int i = Integer.parseInt(st.nextToken());
            int x;

            switch (c) {
                case 1:
                    x = Integer.parseInt(st.nextToken());
                    trains[i - 1] |= 1 << (x - 1);
                    break;
                case 2:
                    x = Integer.parseInt(st.nextToken());
                    trains[i - 1] &= ~(1 << (x - 1));
                    break;
                case 3:
                    trains[i - 1] <<= 1;
                    break;
                case 4:
                    trains[i - 1] >>= 1;
                    break;
            }

            trains[i - 1] &= mask;
        }

        System.out.println(Arrays.stream(trains).distinct().count());
    }
}
